### PR TITLE
Set yanked_info_checksum on gem yank

### PR DESCRIPTION
**Pull request by sonalkr132**
closes: #1382

### CODE REVIEW

File: app/models/deletion.rb [hunk 1]
```diff
@@ -30,7 +30,7 @@ def record_metadata
   end
 
   def expire_api_memcached
-    ["deps/v1/#{rubygem}", "info/#{rubygem}", "versions", "names"].each do |key|
+    ["deps/v1/#{rubygem}", "versions", "names"].each do |key|
```
Comment (segiddins) [inline app/models/deletion.rb]: Why this change?

Comment (sonalkr132) [inline app/models/deletion.rb]: On `gem yank`, info cache is cleared [here](https://github.com/rubygems/rubygems.org/pull/1384/files#diff-175d93318045eb4395c87409e05dc6cfR57). No need to do it again.

Comment (segiddins) [inline app/models/deletion.rb]: but that's before it's committed, so couldn't there be a race where fastly re-requests the page before the commit is done?

Comment (sonalkr132) [inline app/models/deletion.rb]: I believe in the rare condition the `yanked_info_checksum` will be set from cache. I hope this what you are pointing it:
```
* gem yank is requested for `foo`         |
* `info/foo` cache gets be deleted        |
                                          |  * client requests `info/foo`
                                          |  * `info/foo` gets calculated from DB using the state
                                          |     before `gem yank` (since `destroy` is running in transaction)
                                          |     and cache is set
* `rubygem.compact_index_info` gets       |
  called, however now it will end up      |
  reading from cache                      |
```

Comment (segiddins) [inline app/models/deletion.rb]: where is it that the cache gets set? `set_yanked_info_checksum` only sets the `yanekd_info_checksum` in the DB, which won't be used since the version won't be fully yanked yet

Comment (indirect) [inline app/models/deletion.rb]: I would prefer that all cache expiration stay together in the `after_create` method, so that this kind of race condition is not possible even theoretically. 👍

Comment (sonalkr132) [inline app/models/deletion.rb]: @segiddins 
> where is it that the cache gets set?

cache gets set on `GET` request for `info/foo`. Since I am explaining it now, it is most likely thats not what you meant by **fastly re-requests the page**. Can you please explain?

Comment (segiddins) [inline app/models/deletion.rb]: right, when the page gets requested before the transaction that sets the yank is committed, the gem won't show as yanked yet, so fastly will then re-cache the page with the unyanked version

Comment (sonalkr132) [inline app/models/deletion.rb]: > fastly will then re-cache the page with the unyanked version

and yanked_info_checksum will end up with incorrect checksum. Does that count as race condition?

PS: This PR increases number of callbacks on deletion to 5. It feels one too many. I was going to open an issue and work on it separately. May be we can consider it here itself?

**Comment by sonalkr132** [comment]
I have restored purging of `info/foo` in `after_commit`. As @segiddins pointed out, it deals with situation where `info/foo` cache gets set while `deletion.create` transaction is running.

**Comment by indirect** [comment]
I am r+. @segoddins r?

**Comment by indirect** [comment]
I am 👍 on this. @segiddins r?

**Comment by indirect** [comment]
@sonalkr132 once this is deployed, we will need to recalculate every info checksum and every yanked info checksum. Could you write a script to do this that we can run it later?

**Comment by sonalkr132** [comment]
> we will need to recalculate every info checksum and every yanked info checksum

Other alternative is that we regenerate `versions.list`. Current one is 8 months old(delays response time). With a new file we won't have to recalculate `info_checksum` or `yanked_info_checksum` as the file will only have indexed versions. Any appended data to the new file also be correct because after this PR gets deployed we will have `yanked_info_checksum` of yanked versions. Clients with new file will be able to correctly update their local `versions.list` cache.

Also, backfill for checksum takes days to finish.

I hope I got everything right :sweat_smile:

**Comment by indirect** [comment]
@sonalkr132 oh, good point! lets delete all the incorrect checksums and regenerate versions.list one time after this PR is deployed to production. 😅

**Comment by sonalkr132** [comment]
> lets delete all the incorrect checksums

AFAIK, there are no incorrect `info_checksum`(s). [`info_checksum` is set only on `gem push`](https://github.com/rubygems/rubygems.org/blob/master/app/models/pusher.rb#L172). Calculation of `info_checksum` [doesn't use any info related to yanked version](https://github.com/rubygems/rubygems.org/blob/master/app/models/gem_info.rb#L90) either.
We only need to backfill `yanked_info_checksum` for non indexed version.

**Comment by indirect** [comment]
I see... I guess only bundler-api set bad checksums on yanking then 😅
